### PR TITLE
Synchronize Metrics Events

### PR DESCRIPTION
**Pull request by tarekgh**
This change is to synchronize the Metrics events: 

    - Instrument.Publish 
    - MeterListener.Start
    - MeterListener.EnableMeasurementEvents
    - MeterListener.DisableMeasurementEvents
    - MeterListener.Dispose
    - Meter.Dispose
    - Meter.Constructor

This will ensure we'll not send the same notification to the listeners more than one time. Also, will ensure all listener notification (e.g., measurements completed event) will reliably occur.

### CODE REVIEW

**Review by noahfalk — APPROVED** [review]
Looks good other than the potential double publish notification issue

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs [hunk 1]
```diff
@@ -112,16 +110,21 @@ internal static void ValidateTypeParameter<T>()
         }
 
         // Called from MeterListener.EnableMeasurementEvents
-        internal void EnableMeasurement(ListenerSubscription subscription)
+        internal List<object?>? EnableMeasurement(ListenerSubscription subscription)
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs]: I fixed that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs [hunk 2]
```diff
@@ -165,34 +191,94 @@ public void RecordObservableInstruments()
...
+        internal static void Publish(Instrument instrument)
+        {
+            List<MeterListener>? listeners = null;
+
+            lock (Instrument.SyncObject)
             {
-                object? state = current.Value.DisableMeasurements(this);
-                MeasurementsCompleted?.Invoke(current.Value, state);
-                current = current.Next;
+                if (instrument is null || instrument.Meter.Disposed || !instrument.Meter.AddInstrument(instrument))
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs]: Good point. The instrument cannot be null here as it is called from Instrument.Publish. I fixed that.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs [hunk 3]
```diff
@@ -165,34 +191,94 @@ public void RecordObservableInstruments()
...
+
+                LinkedListNode<MeterListener>? current = s_allStartedListeners.First;
+                if (current is not null)
+                {
+                    listeners = new List<MeterListener>();
+
+                    do
+                    {
+                        // Notify only if the listener didn't already registered this instrument before as the notification should be already happened.
+                        if (!current.Value.IsRegisteredInstrument(instrument))
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs]: I applied your suggestion. I agree it is simpler now than before.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs [hunk 4]
```diff
@@ -19,6 +19,8 @@ public abstract class Instrument
         internal static KeyValuePair<string, object?>[] EmptyTags { get; } = Array.Empty<KeyValuePair<string, object?>>();
 #endif // NO_ARRAY_EMPTY_SUPPORT
 
+        internal static object SyncObject { get; } = new object();
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs:30]: I address this feedback too.

**Pull request merged**